From 10e0afd044c64d3ff3a9dcf23eb98597d76c1d82 Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 24 Oct 2014 20:28:18 -0400
Subject: [net] openvswitch: restore OVS_FLOW_CMD_NEW notifications

Message-id: <9ee54b2b4908f596bc3b161c5c5bd5de36eb3ae0.1414182116.git.jbenc@redhat.com>
Patchwork-id: 98446
O-Subject: [RHEL7.1 net PATCH 10/11] openvswitch: restore OVS_FLOW_CMD_NEW notifications
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8159383

commit 9b67aa4a82492f128adfccc63e61ab57c1ce1dfd
Author: Samuel Gauthier <samuel.gauthier@6wind.com>
Date:   Thu Sep 18 10:31:04 2014 +0200

    openvswitch: restore OVS_FLOW_CMD_NEW notifications

    Since commit fb5d1e9e127a ("openvswitch: Build flow cmd netlink reply only if needed."),
    the new flows are not notified to the listeners of OVS_FLOW_MCGROUP.

    This commit fixes the problem by using the genl function, ie
    genl_has_listerners() instead of netlink_has_listeners().

    Signed-off-by: Samuel Gauthier <samuel.gauthier@6wind.com>
    Signed-off-by: Nicolas Dichtel <nicolas.dichtel@6wind.com>
    Acked-by: Pravin B Shelar <pshelar@nicira.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/datapath.c b/net/openvswitch/datapath.c
index 4e48e0b..458f94c 100644
--- a/net/openvswitch/datapath.c
+++ b/net/openvswitch/datapath.c
@@ -78,11 +78,12 @@ static const struct genl_multicast_group ovs_dp_vport_multicast_group = {
 
 /* Check if need to build a reply message.
  * OVS userspace sets the NLM_F_ECHO flag if it needs the reply. */
-static bool ovs_must_notify(struct genl_info *info,
-       const struct genl_multicast_group *grp)
+static bool ovs_must_notify(struct genl_family *family, struct genl_info *info,
+       unsigned int group)
 {
  return info->nlhdr->nlmsg_flags & NLM_F_ECHO ||
-  netlink_has_listeners(genl_info_net(info)->genl_sock, 0);
+        genl_has_listeners(family, genl_info_net(info)->genl_sock,
+      group);
 }
 
 static void ovs_notify(struct genl_family *family,
@@ -764,7 +765,7 @@ static struct sk_buff *ovs_flow_cmd_alloc_info(const struct sw_flow_actions *act
 {
  struct sk_buff *skb;
 
- if (!always && !ovs_must_notify(info, &ovs_dp_flow_multicast_group))
+ if (!always && !ovs_must_notify(&dp_flow_genl_family, info, 0))
   return NULL;
 
  skb = genlmsg_new_unicast(ovs_flow_cmd_msg_size(acts), info, GFP_KERNEL);
-- 
1.7.1